ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger/Query.hs

```diff
@@ -309,14 +326,19 @@ answerQueryAnytime HardForkLedgerConfig{..} =
 -------------------------------------------------------------------------------}
 
 data QueryHardFork xs result where
-  GetInterpreter :: QueryHardFork xs (History.Interpreter xs)
-  GetCurrentEra  :: QueryHardFork xs (EraIndex xs)
+  GetInterpreter  :: QueryHardFork xs (History.Interpreter xs)
+  GetCurrentEra   :: QueryHardFork xs (EraIndex xs)
+  GetLedgerCfg    :: QueryHardFork xs (HardForkLedgerConfig xs)
```

Comment:
It feels like I should have thought of this sooner... sorry.

The "give me the ledger config" query actually seems orthogonal to HFC. It just happens that HFC is the outermost layer of the blocks on `mainnet`. But that is an implementation detail, not necessarily true for other (future/anticipated) uses of this library.

Should we instead introduce an outer layer of query that would host the generic queries you're working to add? They seem focused on data that is just a consequence of the architecture, largely irrespective of the block type (other than the data varying with the block type).

Thoughts @DavidEichmann @dcoutts @EncodePanda ?

Reply:
I'm a bit unsure what you are suggesting @nfrisby. It seems to me that the essence of the HFC is to abstract over a type level list of eras (here `xs`). I utilize this in all the queries I plan on adding, either in the return type or in the query parameters. For example, when querying the slot length, the query still requires a `EraIndex xs` even though the resulting `SlotLength` type is era agnostic (you mentioned this in your comment). Are you suggesting we can change the queries to *not* be parameterized over `xs`? I suppose we could use `Int` instead of `EraIndex xs`, but we'd lose some type safety and I don't see what we gain from that. In the case of this PR, the query result must be parameterized over the eras, else what would the query return? So it seems to me we need to parameterize of `xs` and that is the essence of the HFC, so it makes sense that these queries are associated with the HFC i.e. placed in `Query (HardForkBlock xs)`. Is there something I'm missing? Do you have a more concrete example of what you're thinking of?

Reply:
> the essence of the HFC is to abstract over a type level list of eras (here xs). I utilize this in all the queries I plan on adding

But the result type of `HardForkLedgerConfig xs` is the same as `LedgerConfig blk`, under the assumption (true eg on mainnet) that `blk ~ HardForkBlock xs`, isn't' it?

> In the case of this PR, the query result must be parameterized over the eras, else what would the query return?

I think the "give me the ledger config" query could instead inhabit the type `Query blk (LedgerConfig blk)` if we were to reorganize things a bit. That query's core semantics make sense for any `blk` type, regardless of whether it's headed by `HardForkBlock`.

Or am I missing something?

I chatted with Edsko and he reminded me that Duncan and he were basically suggesting that every block type going forward should be an application of `HardForkBlock`. If we really commit to that , then my observation here is rather moot. Except... note the comment on `QueryHardFork`:

https://github.com/input-output-hk/ouroboros-network/blob/5916d85bd37246fbbea3f080559f00d015f33071/ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Ledger/Query.hs#L106-L113

Thus a chain running a block type `blk ~ HardForkBlock [TheOneBlock]` will not be able to use the `GetLedgerConfig` query that you're introducing in this PR. It seems to me like such a chain should be able to (see all the above reasoning).

Does that all clarify my concern here?

Reply:
Thanks for your response, @nfrisby. This is making more sense to me now. I think you are suggesting a top level query type like this:

```haskell
data TopQuery blk result where
  QueryLedgerConfig :: TopQuery blk (LedgerConfg blk)
  InnerQuery :: Query blk result -> TopQuery blk result
```

Another interpretation of `TopQuery` is that we're specifying a set of common queries that all block types support. Then we'd have to change the `answerQuery` function in the `QueryLedger` class to use `TopQuery`:

```haskell
class (ShowQuery (Query blk), SameDepIndex (Query blk)) => QueryLedger blk where
  answerQuery :: ExtLedgerCfg blk -> Query blk result -> ExtLedgerState blk -> result
```

Reply:
Yes, that's certainly the spirit of what I had in mind. Would you discuss this path -- specifically, whether it seems worthwhile -- with Duncan? (I'm guessing you're chatting with some regularity?) In particular, my observation about the constraint on `QueryHardFork` seems more important than I first had guessed.

Reply:
FYI, Edsko pointed out that this module exists, which lifts the "is multiple eras" constraint.

https://github.com/input-output-hk/ouroboros-network/blob/4404be28dc53948eac7f47f116d2b01c128016b1/ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Compat.hs

Reply:
Thoughts:
* Yes we do expect all uses of consensus to use the HFC, even if it's for a single era. If (as I assume it does) simplify things, then that's all to the better.
* The `IsNonEmpty xs => (x ': xs)` constraint is something I have never _really_ understood. My limited understanding is that it helped or helps with enforcing and testing that standalone eras and singleton-HFC-wrapped eras are equivalent. This was practically useful when we had the single Byron era actually in use. In practice now we don't need that, but do we still need it to help with the simplicity of test properties? I do not know. If we don't need the constraint for tests I'd be happy to drop it.
* Some of the queries that David is adding just depend on the block type (where for the HFC that is ~ [...] to some list of eras), but some also fit the pattern of the other HFC-query classes, of per-era, rather than era-independent. So given that, it's not clear to me that there's a big advantage to be had by pulling the era-independent ones out to a new top level wrapper.

Does that help? We can have a more detailed chat if needed.

Reply:
Ok. So what I'm getting out of this is that really, instead of creating a new "top level query" type, `TopQuery blk result`,  we accept that HFC is hear to stay and hence such a top level query type already exists: `data instance Query (HardForkBlock xs) result`. We should perhaps move the `GetLedgerConfig` constructor from `data QueryHardFork` to `data instance Query (HardForkBlock xs)`:

```haskell
data instance Query (HardForkBlock xs) :: Type -> Type where
  -- ...
  GetLedgerConfig :: Query (HardForkBlock xs) (LedgerConfg (HardForkBlock xs))
```
In the case that `xs ~ [x]` (i.e. is a single era), and you just want `LedgerConfig x`, it's simply a matter of unwrapping it from the `HardForkLedgerConfig` and possibly converting from a `PartialLedgerConfig x` via `completeLedgerConfig`. Does this make sense @dcoutts @nfrisby ?

Reply:
Yes, thanks both; that all makes sense.

I'll study a bit and also ask @edsko tomorrow for his thoughts on lifting the restriction. (Suppose he argues we should keep it; in that case, I imagine his objections would also apply to putting `GetLedgerConfig` in `Query` instead of in `QueryHardFork`, since that would also spoil the isomorphism.)

FWIW, it looks like `cardano-node` does indeed use the `CompatQuery` interface (which lifts the restriction). However, I don't yet know _why_ they do that. This is just another thread for one of us to eventually pull on.

```
~/cardano-node$ git grep -l -e CompatQuery origin/master
origin/master:cardano-node/src/Cardano/Node/Query.hs
```

Reply:
Having cogitated on this a bit more, I'm more sympathetic to the general top level query idea from @nfrisby. I'd sumarrise the idea as saying that these are queries that consensus can _always_ support, for any protocol/ledger, irrespective of whether that protocol is composite (HFC) or not.

For the `QueryHardFork` queries, existing, or the new one, or other future ones:
* `GetInterpreter` this is indeed HFC-specific, the notion of the interpreter only makes sense for composite protocols
* `GetCurrentEra` this is clearly HFC-specific
* `GetLedgerConfig` this is indeed not HFC-specific. It would make sense for any block type, composite or not.
* A future "get initial ledger state". This also need not be HFC specific. And possibly the type would be simpler if it were not HFC specific. For the HFC case it'd return a ledger state in the union type, rather than specifically in the type of the first era.

So on reflection it's actually quite a reasonable idea to say we should hoist these to the top. In practice I'm not quite sure how best to do that. The `Query` type is a data family of course, so we don't currently have a convenient place to stick in some extra constructors. I suppose it'd need a distinction between the top level query type, with constructors for the generic queries, and the protocol-specific queries, and the data family for the types of those protocol-specific queries.

Reply:
> I suppose it'd need a distinction between the top level query type, with constructors for the generic queries, and the protocol-specific queries, and the data family for the types of those protocol-specific queries.

That'd be my first choice, I think. We might have to do a bit of a dance regarding backward compatible serializers for that route. But it seems the cleanest in the end.

A looser alternative is to require that any `Query` instance satisfies a type class that ensures the general queries can be constructed. This would probably lead to each ledger having to reimplement the generic queries redundantly; but the generic queries are likely not very complicated, so maybe this is a reasonable at-least-temporary approach (which will probably have an easier time satisfying the backwards compatibility).

Reply:
> `GetInterpreter` this is indeed HFC-specific,

This is just a reactionary thought at the moment: we could return a degenerate `Interpreter` for a non-HFC block, couldn't we? In particular, the `HardForkIndices blk :: [*]` type family is applicable to any block type. EG we already have `type instance HardForkIndices ByronBlock = '[ByronBlock]`

Reply:
Update: We had a call. We're going to assess how disruptive it would be to introduce the new top-level query type that contains the black-independent queries. EG Several of the planned queries would fit naturally there, so it seems worth it. Unless we find unexpected prereqs that we don't want to pay down now.